drivers/serial/uart_acts.c

```diff
@@ -0,0 +1,501 @@
+
+	if (uart->ctrl & (0x1 << UART_CTL_RXIE))
+		return (uart->stat & (0x1 << UART_STA_RIP));
+	else
+		return 0;
+}
+
+/** Interrupt driven pending status function */
+static int uart_acts_irq_is_pending(struct device *dev)
+{
+	return (uart_acts_irq_tx_ready(dev) || uart_acts_irq_rx_ready(dev));
+}
+
+/** Interrupt driven interrupt update function */
+static int uart_acts_irq_update(struct device *dev)
+{
+	return 1;
+}
+
+/** Set the callback function */
+static void uart_acts_irq_callback_set(struct device *dev,
+				       uart_irq_callback_user_data_t cb,
+				       void *cb_data)
+{
+	struct uart_acts_dev_data * const dev_data = DEV_DATA(dev);
+
+	dev_data->cb = cb;
+	dev_data->cb_data = cb_data;
+}
+
+/**
+ * @brief Interrupt service routine.
+ *
+ * This simply calls the callback function, if one exists.
+ *
+ * @param arg Argument to ISR.
+ *
+ * @return N/A
+ */
+void uart_acts_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct uart_acts_dev_data * const dev_data = DEV_DATA(dev);
+
+	if (dev_data->cb) {
+		dev_data->cb(dev_data->cb_data);
+	}
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api uart_acts_driver_api = {
+	.poll_in          = uart_acts_poll_in,
+	.poll_out         = uart_acts_poll_out,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill        = uart_acts_fifo_fill,
+	.fifo_read        = uart_acts_fifo_read,
+	.irq_tx_enable    = uart_acts_irq_tx_enable,
+	.irq_tx_disable   = uart_acts_irq_tx_disable,
+	.irq_tx_ready     = uart_acts_irq_tx_ready,
+	.irq_rx_enable    = uart_acts_irq_rx_enable,
+	.irq_rx_disable   = uart_acts_irq_rx_disable,
+	.irq_tx_complete  = uart_acts_irq_tx_complete,
+	.irq_rx_ready     = uart_acts_irq_rx_ready,
+	.irq_is_pending   = uart_acts_irq_is_pending,
+	.irq_update       = uart_acts_irq_update,
+	.irq_callback_set = uart_acts_irq_callback_set,
+#endif
+};
+
+/**
+ * @brief Initialize UART channel
+ *
+ * This routine is called to reset the chip in a quiescent state.
+ * It is assumed that this function is called only once per UART.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0 on success
+ */
+int uart_acts_init(struct device *dev)
+{
+	volatile struct acts_uart_controller *uart = UART_STRUCT(dev);
+	struct uart_acts_dev_data * const dev_data = DEV_DATA(dev);
+	struct device *clock_dev;
+	u32_t clock_freq;
+
+	int err;
+
+	/* reset uart */
+	acts_reset_peripheral(dev_data->reset_id);
+
+	/* enable uart clock */
+	clock_dev = device_get_binding(dev_data->clock_name);
+	if (!clock_dev) {
+		return -EINVAL;
+	}
+
+	clock_control_on(clock_dev, dev_data->clock_subsys);
+
+	if (clock_control_get_rate(clock_dev, dev_data->clock_subsys,
+				   &clock_freq)) {
+		return -EINVAL;
+	}
+
+	dev_data->clock_freq = clock_freq;
+
+	/* Set baud rate */
+	err = baudrate_set(dev, dev_data->baud_rate);
+	if (err)
+		return err;
+
+	/* Enable receiver and transmitter */
+	uart->ctrl = ((0x1 << 31) | (0x1 << 30) | (0x1 << 23) | (0x1 << 22) |
```

Comment:
Can this magic number be replaced with defines?

Reply:
Done